### ISSUE TITLE
cluster-autoscaler 1.15.5 does not work with IAM Roles for Service Accounts

### ISSUE THREAD

**Issue by johnjeffers** [kubernetes/autoscaler#2920]
Just upgraded my EKS cluster to 1.15, and upgraded cluster-autoscaler along with it.

It appears that 1.15.5 has the same problem that we used to have with 1.14. It does not work with IAM Roles for Service Accounts.

Getting the following error:

```E0311 03:36:42.052777       1 aws_manager.go:259] Failed to regenerate ASG cache: cannot autodiscover ASGs: AccessDenied: User: arn:aws:sts::<my-account>:assumed-role/eks-worker/i-059d924ed6f52a032 is not authorized to perform: autoscaling:DescribeTags```

This should not be happening, as the `autoscaling:DescribeTags` permission is assigned to the service-account-based IAM Role, not to the `eks-worker` instance profile.

This problem was fixed in [1.14.6](https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-1.14.6) but doesn't seem to have made it into 1.15.5

**Comment by rochacon** [comment]
I hit the same issue, downgrading to 1.14.7 seems to work fine. I believe we need to cherry-pick #2323 (or a more recent AWS SDK update).

**Comment by Jeffwan** [comment]
I will go fix this issue. I remember I did the cherry-pick. Need to double check.

**Comment by johnjeffers** [comment]
@Jeffwan Thank you!!

**Comment by Jeffwan** [comment]
@losipiuk @aleksandra-malinowska Trying to cherry-pick a change to v1.15.  that branch doesn't have `./hack/update-vendor.sh` and  `k8s/kubernetes` dependencies points to `/tmp/abc/kubernetes`. 



Another thing is I notice 

```
# Using go mod
build-binary: clean deps
        $(ENVVAR) GOOS=$(GOOS) go build -o cluster-autoscaler ${TAGS_FLAG}

# Seems godep
build-in-docker: clean docker-builder
        docker run -v `pwd`:/gopath/src/k8s.io/autoscaler/cluster-autoscaler/ autoscaling-builder:latest bash -c 'cd /gopath/src/k8s.io/autoscaler/cluster-autoscaler && BUILD_TAGS=${BUILD_TAGS} make build-binary'
```

How can I manage dependencies in this case?

**Comment by ajohnstone** [comment]
Bumping version should solve #2829 with sts private endpoints too

**Comment by canhnt** [comment]
@Jeffwan: go.mod in  the branch `cluster-autoscaler-release-1.15` does not define specific versions of modules in k8s.io. That makes updating dependencies with `go mod` difficult. One possibility is to assume all k8s.io modules are on branch of k8s-v1.15.x, then run the test after upgrading aws-sdk-go. But then we need to checkout all modules' version an mount `k8s.io` to `/tmp/abc/kubernetes` path.

There is another hacky way is to manually import transitive dependencies (but not indirect) in aws-sdk-go v1.29.29 (latest on 30/03). I tested in [my fork](https://github.com/canhnt/autoscaler/tree/task/upgrade-aws-sdk-go-v1.29.29). The build and test did well without any issue.  The target [Docker image](https://hub.docker.com/repository/docker/canhnt/cluster-autoscaler/tags?page=1) does not have exceptions anymore.  Might it be the possible way?

**Comment by Jeffwan** [comment]
@canhnt I sync with release team. We backport `./hack/update-vendor.sh` from later branch and update base for those changes.. I update to v1.28.14 and problem should be resolved. Thanks for the feeback!

**Comment by Jeffwan** [comment]
For folks who can not wait for upstream release, please either build `cluster-autoscaler-release-1.15` by yourself or use my image `seedjeffwan/cluster-autoscaler:1.15-dev` to mitigate the problem for short term.

**Comment by towca** [comment]
FYI: We'll be doing new patch releases on Monday: #2988.

**Comment by mbarrien** [comment]
Note that a similar patch will be required in the cluster-autoscaler 1.16 branch (since 1.16 is using aws-sdk-go 1.23.12, and IRSA support was introduced in 1.23.13). Not an immediate issue, but will be once AWS release EKS with 1.16; if you're already doing a PR for 1.15, might as well do the same thing for 1.16.

**Comment by Jeffwan** [comment]
@mbarrien Thanks for reminder, yes. I will update to 1.28.x to support regional STS endpoint as well

**Comment by Jeffwan** [comment]
Please use this version https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-1.15.6

**Comment by Jeffwan** [comment]
> Note that a similar patch will be required in the cluster-autoscaler 1.16 branch (since 1.16 is using aws-sdk-go 1.23.12, and IRSA support was introduced in 1.23.13). Not an immediate issue, but will be once AWS releases EKS with 1.16; if you're already doing a PR for 1.15, might as well do the same thing for 1.16.

A PR has been filed to address 1.16 issues for EKS IRSA 
https://github.com/kubernetes/autoscaler/pull/3003

It will be included in next version

**Comment by johnjeffers** [comment]
Thank you @Jeffwan !!!

**Comment by damscott** [comment]
> Please use this version https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-1.15.6

@Jeffwan The image for the 1.15 version isn't in the registry:
```shell
❯ docker pull k8s.gcr.io/cluster-autoscaler:v1.15.5
v1.15.5: Pulling from cluster-autoscaler
07508dcb5d70: Pull complete
57be50ba1432: Pull complete
845bb582d551: Pull complete
7696b1c466cf: Pull complete
Digest: sha256:36438427003b380d7ab3b28fb1818355a36de0f1cc623e7d13b4ac8680e14259
Status: Downloaded newer image for k8s.gcr.io/cluster-autoscaler:v1.15.5
k8s.gcr.io/cluster-autoscaler:v1.15.5

❯ docker pull k8s.gcr.io/cluster-autoscaler:v1.15.6
Error response from daemon: manifest for k8s.gcr.io/cluster-autoscaler:v1.15.6 not found: manifest unknown: Failed to fetch "v1.15.6" from request "/v2/cluster-autoscaler/manifests/v1.15.6".
```

**Comment by mbarrien** [comment]
Use one of the official registry URLs listed at the bottom of the release notes. There was a reorganization of the registry locations.

**Comment by damscott** [comment]
> Use one of the official registry URLs listed at the bottom of the release notes. There was a reorganization of the registry locations.

Thanks, that worked. I'll put in a PR on the cluster-autoscaler helm chart, which [is still pointing at the old URL](https://github.com/helm/charts/blob/master/stable/cluster-autoscaler/values.yaml#L55).

**Comment by skadem07** [comment]
@johnjeffers Hi, i'm facing the same issue, can you pls share me your yaml file, I'm not able to figure out where i'm wrong. Thanks!

**Comment by allamand** [comment]
Hi, 

I have the same problem with eu.gcr.io/k8s-artifacts-prod/autoscaling/cluster-autoscaler:v1.18.2 while it works with eu.gcr.io/k8s-artifacts-prod/autoscaling/cluster-autoscaler:v1.18.1

**Comment by johnjeffers** [comment]
@allamand are you running cluster-autoscaler 1.18 in EKS?

**Comment by allamand** [comment]
@johnjeffers yes, but after switching several times it seems to also works with the .2 version.. maybe a problem with my IRSA config..

**Comment by johnjeffers** [comment]
@allamand If you are on EKS, you should not be using 1.18.2. cluster-autoscaler docs say to use the version that corresponds to your Kubernetes version, and the latest Kubernetes for EKS is 1.17. So, for example, if your EKS is on Kubernetes 1.16, you should use cluster-autoscaler 1.16.x.

**Comment by Jeffwan** [comment]
We are scoping 1.18 recently and it will be available soon. currently , there's no support.  User has to choose corresponding minor version with largest patch version. CA uses scheduler logic to do simulation. Keep in mind this is important!

**Comment by vainkop** [comment]
I'm using `k8s.gcr.io/autoscaling/cluster-autoscaler:v1.20.0` on EKS 1.20 & have the same problem as @johnjeffers had in the very beginning:
```
0609 23:36:00.590745 1 aws_manager.go:265] Failed to regenerate ASG cache: cannot autodiscover ASGs: AccessDenied: User: arn:aws:sts::1234567890123:assumed-role/dev-eks-spot-gwNaBAcu20210609220001136000000009/i-089aef7c05313c957 is not authorized to perform: autoscaling:DescribeTags
status code: 403, request id: 2f4de852-ad1d-43fc-8753-d61912fb5ed9
F0609 23:36:00.590781 1 aws_cloud_provider.go:382] Failed to create AWS Manager: cannot autodiscover ASGs: AccessDenied: User: arn:aws:sts::1234567890123:assumed-role/dev-eks-spot-gwNaBAcu20210609220001136000000009/i-089aef7c05313c957 is not authorized to perform: autoscaling:DescribeTags
status code: 403, request id: 2f4de852-ad1d-43fc-8753-d61912fb5ed9
```

Here's how I deploy the cluster-autoscaler:

$ cat cluster-autoscaler-iam.json
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": [
        "autoscaling:DescribeAutoScalingGroups",
        "autoscaling:DescribeAutoScalingInstances",
        "autoscaling:DescribeLaunchConfigurations",
        "autoscaling:DescribeTags",
        "autoscaling:SetDesiredCapacity",
        "autoscaling:TerminateInstanceInAutoScalingGroup",
        "ec2:DescribeLaunchTemplateVersions"
      ],
      "Resource": "*",
      "Effect": "Allow"
    }
  ]
}
```
$ cat template-trust.json
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "arn:aws:iam::$AWS_ACCOUNT_ID:oidc-provider/$OIDC_PROVIDER"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringEquals": {
          "$OIDC_PROVIDER:sub": "system:serviceaccount:kube-system:cluster-autoscaler"
        }
      }
    }
  ]
}

```
$ cat ServiceAccount.yml
```
---
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app.kubernetes.io/name: cluster-autoscaler
  name: cluster-autoscaler
  namespace: kube-system
---
```

```
export CLUSTER_NAME=eks-test
export AWS_DEFAULT_REGION=us-east-1

aws iam create-policy \
  --policy-name ClusterAutoscalerIAM_CA_Policy \
  --policy-document file://cluster-autoscaler-iam.json

export ClusterAutoscalerIAM_CA_Policy_ARN=$(aws iam list-policies --query 'Policies[?PolicyName==`ClusterAutoscalerIAM_CA_Policy`].Arn' --output text)

kubectl apply -f ServiceAccount.yml

export AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query "Account" --output text)
export OIDC_PROVIDER=$(aws eks describe-cluster --name $CLUSTER_NAME --query "cluster.identity.oidc.issuer" --output text | sed -e "s/^https:\/\///")

envsubst < template-trust.json > trust.json

aws iam create-role \
  --role-name $CLUSTER_NAME-cluster-autoscaler \
  --assume-role-policy-document file://trust.json \
  --description "$CLUSTER_NAME-cluster-autoscaler role"

aws iam attach-role-policy --role-name $CLUSTER_NAME-cluster-autoscaler --policy-arn=$ClusterAutoscalerIAM_CA_Policy_ARN

export ROLE_ARN=$(aws iam get-role --role-name $CLUSTER_NAME-cluster-autoscaler | jq -c '.[]' | jq -r '.Arn')

kubectl annotate --overwrite serviceaccount -n kube-system cluster-autoscaler eks.amazonaws.com/role-arn=$ROLE_ARN

helm repo add autoscaler https://kubernetes.github.io/autoscaler && helm repo update

helm install cluster-autoscaler autoscaler/cluster-autoscaler --set 'autoDiscovery.clusterName'=$CLUSTER_NAME --set awsRegion="$AWS_DEFAULT_REGION"
```

**Comment by ekamdz** [comment]
@Jeffwan , Is it available for 1.18 Cluster autoscaler?

I am using image **_k8s.gcr.io/autoscaling/cluster-autoscaler:v1.18.3_**

Getting this error:-
```
E0730 14:18:49.622795       1 aws_manager.go:259] Failed to regenerate ASG cache: cannot autodiscover ASGs: WebIdentityErr: failed to retrieve credentials
caused by: AccessDenied: Not authorized to perform sts:AssumeRoleWithWebIdentity
	status code: 403, request id: c3d5fdbd-fb4e-4e6f-a329-b687f8ccc234
F0730 14:18:49.622817       1 aws_cloud_provider.go:379] Failed to create AWS Manager: cannot autodiscover ASGs: WebIdentityErr: failed to retrieve credentials
caused by: AccessDenied: Not authorized to perform sts:AssumeRoleWithWebIdentity
	status code: 403, request id: c3d5fdbd-fb4e-4e6f-a329-b687f8ccc234

```
